pkg/dockerregistry/server/metrics/metrics.go

```diff
@@ -1,83 +1,188 @@
+type Pullthrough interface {
+	// RepositoryRetriever wraps RepositoryRetriever to collect statistics.
+	RepositoryRetriever(retriever registryclient.RepositoryRetriever) registryclient.RepositoryRetriever
+
+	// DigestBlobStoreCache() returns an interface to count cache hits/misses
+	// for pullthrough blobstores.
+	DigestBlobStoreCache() Cache
+}
+
+// Storage is a ste of metrics for the storage subsystem.
+type Storage interface {
+	// StorageDriver wraps distribution/registry/storage/driver.StorageDriver
+	// to collect statistics.
+	StorageDriver(driver storagedriver.StorageDriver) storagedriver.StorageDriver
+}
+
+func dockerErrorCode(err error) string {
+	if e, ok := err.(errcode.Error); ok {
+		return e.ErrorCode().String()
+	}
+	return "UNKNOWN"
+}
+
+func pullthroughRepositoryWrapper(ctx context.Context, sink Sink, registry string, funcname string, f func(ctx context.Context) error) error {
+	registry = strings.ToLower(registry)
+	defer NewTimer(sink.PullthroughRepositoryDuration(registry, funcname)).Stop()
+	err := f(ctx)
+	if err != nil {
+		sink.PullthroughRepositoryErrors(registry, funcname, dockerErrorCode(err)).Inc()
+	}
+	return err
+}
 
-// Register the metrics.
-func Register() {
-	registryAPIRequests = prometheus.NewHistogramVec(
-		prometheus.HistogramOpts{
-			Namespace: registryNamespace,
-			Subsystem: registrySubsystem,
-			Name:      "request_duration_seconds",
-			Help:      "Request latency summary in microseconds for each operation",
-		},
-		[]string{"operation", "name"},
-	)
-	prometheus.MustRegister(registryAPIRequests)
-}
-
-// Timer is a helper type to time functions.
-type Timer interface {
-	// Stop records the duration passed since the Timer was created with NewTimer.
-	Stop()
-}
-
-// NewTimer wraps the HistogramVec and used to track amount of time passed since the Timer was created.
-func NewTimer(collector *prometheus.HistogramVec, labels []string) Timer {
-	return &metricTimer{
-		collector: collector,
-		labels:    labels,
-		startTime: time.Now(),
+type repositoryRetriever struct {
+	retriever registryclient.RepositoryRetriever
+	sink      Sink
+}
+
+func (rr repositoryRetriever) Repository(ctx gocontext.Context, registry *url.URL, repoName string, insecure bool) (repo distribution.Repository, err error) {
+	err = pullthroughRepositoryWrapper(ctx, rr.sink, registry.Host, "Init", func(ctx context.Context) error {
+		repo, err = rr.retriever.Repository(ctx, registry, repoName, insecure)
+		return err
+	})
+	if err != nil {
+		return repo, err
 	}
+	return wrapped.NewRepository(repo, func(ctx context.Context, funcname string, f func(ctx context.Context) error) error {
+		return pullthroughRepositoryWrapper(ctx, rr.sink, registry.Host, funcname, f)
+	}), nil
 }
 
-type metricTimer struct {
-	collector *prometheus.HistogramVec
-	labels    []string
-	startTime time.Time
+func storageErrorCode(err error) string {
+	switch err.(type) {
+	case storagedriver.ErrUnsupportedMethod:
+		return "UNSUPPORTED_METHOD"
```

Comment:
Should we ignore this errors?

Reply:
why would we?  it seems like if we get this it would be really bad and thus a metric state worth alerting on?

Reply:
This error indicates that the storage doesn't support request offloading. The file system storage is one such and it always returns this error.

An example of the storage that supports this mechanism is the S3 storage, it can redirect clients directly to S3, so we don't need to proxy blobs through ourselves.

Another error that I'm going to ignore is `io.EOF`.

Reply:
got it..so it's not a fatal error.